src/protocols/unreal2/types.rs

```diff
@@ -0,0 +1,256 @@
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct MutatorsAndRules {
+    pub mutators: HashSet<String>,
+    pub rules: HashMap<String, Vec<String>>,
+}
+
+impl MutatorsAndRules {
+    pub fn parse<B: ByteOrder>(&mut self, buffer: &mut Buffer<B>) -> GDResult<()> {
+        while buffer.remaining_length() > 0 {
+            let key = buffer.read_string::<Unreal2StringDecoder>(None)?;
+            let value = buffer.read_string::<Unreal2StringDecoder>(None).ok();
+
+            if key.eq_ignore_ascii_case("mutator") {
+                if let Some(value) = value {
+                    self.mutators.insert(value);
+                }
+            } else {
+                let rule_vec = self.rules.get_mut(&key);
+
+                let rule_vec = if let Some(rule_vec) = rule_vec {
+                    rule_vec
+                } else {
+                    self.rules.insert(key.clone(), Vec::default());
+                    self.rules
+                        .get_mut(&key)
+                        .expect("Value should be in HashMap after we inserted")
+                };
+
+                if let Some(value) = value {
+                    rule_vec.push(value);
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+/// Unreal 2 players and bots.
+#[derive(Clone, Debug, Default, PartialEq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct Players {
+    /// List of players returned by server (without 0 ping).
+    pub players: Vec<Player>,
+    /// List of bots returned by server (players with 0 ping).
+    pub bots: Vec<Player>,
+}
+
+impl Players {
+    /// Pre-allocate the vectors inside the players struct based on the provided
+    /// capacity.
+    pub fn with_capacity(capacity: usize) -> Self {
+        Players {
+            players: Vec::with_capacity(capacity),
+            // Allocate half as many bots as we don't expect there to be as many
+            bots: Vec::with_capacity(capacity / 2),
+        }
+    }
+
+    /// Parse a raw buffer of players into the current struct.
+    pub fn parse<B: ByteOrder>(&mut self, buffer: &mut Buffer<B>) -> GDResult<()> {
+        while buffer.remaining_length() > 0 {
+            let player = Player {
+                id: buffer.read()?,
+                name: buffer.read_string::<Unreal2StringDecoder>(None)?,
+                ping: buffer.read()?,
+                score: buffer.read()?,
+                stats_id: buffer.read()?,
+            };
+
+            // If ping is 0 the player is a bot
+            if player.ping == 0 {
+                self.bots.push(player);
+            } else {
+                self.players.push(player);
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Length of both players and bots.
+    pub fn total_len(&self) -> usize { self.players.len() + self.bots.len() }
+}
+
+/// Unreal 2 player info.
+#[derive(Clone, Debug, PartialEq)]
```

Comment:
Make sure we have all the standard derives (`Debug, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord`) on all types (if they can be added).

Reply:
I think I got all of them in: https://github.com/gamedig/rust-gamedig/pull/124/commits/88b3cfa27d149ea27ab63dbaa56e9ed2424db66d